### PR TITLE
Implement new begin blocker interface

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #4559

### Commit Message / Changelog Entry

N/A part of SDK upgrade

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by chatton** [comment]
The original issue mentioned a test for stopping/starting a chain in a test, I'm not sure I understand the additional benefit that we get from having that. I think this addition to the localhost test should ensure our begin blocker is correctly wired.

Let me know if I'm missing something @colin-axner !

File: e2e/tests/transfer/localhost_test.go [hunk 1]
```diff
@@ -50,6 +50,18 @@ func (s *LocalhostTransferTestSuite) TestMsgTransfer_Localhost() {
...
+	t.Run("verify begin blocker was executed", func(t *testing.T) {
+		cs, err := s.QueryClientState(ctx, chainA, exported.LocalhostClientID)
+		s.Require().NoError(err)
+		originalHeight := cs.GetLatestHeight()
+
+		s.Require().NoError(test.WaitForBlocks(ctx, 1, chainA), "failed to wait for blocks")
+
+		cs, err = s.QueryClientState(ctx, chainA, exported.LocalhostClientID)
+		s.Require().NoError(err)
+		s.Require().True(cs.GetLatestHeight().GT(originalHeight), "client state height was not incremented")
```
Comment (colin-axner) [inline e2e/tests/transfer/localhost_test.go:62]: we could also check that latest height == ctx height

**Review by colin-axner — APPROVED** [review]
Thanks @chatton!

**Comment by colin-axner** [comment]
> The original issue mentioned a test for stopping/starting a chain in a test, I'm not sure I understand the additional benefit that we get from having that.

It would be to detect the missing begin blocker for the capability module. If you stop and start all nodes, you reset the mem store, if the begin block is missing from capability, the mem store will not be reinitialized and any IBC action afterward, such as sending a packet will fail. 

In the short term, I think it makes sense to either add a test or add the stop start to the genesis test or IBC transfer. In the long term, we could consider adding randomness into our testing (allowing random tests to be stop/started)

File: e2e/tests/transfer/localhost_test.go (hunk 1 above)
Comment (chatton) [reply e2e/tests/transfer/localhost_test.go:62]: do you mean `ctx`? We we don't have access to the height the way we do in the unit tests, let me know if I'm misunderstanding the suggestion!

Comment (colin-axner) [reply e2e/tests/transfer/localhost_test.go:62]: This is true, I guess you'd have to query for the node status and pull the latest height from there. I'm totally happy with the current approach, feel free to resolve

Comment (chatton) [reply e2e/tests/transfer/localhost_test.go:62]: I think at the moment this test will always fail if begin blocker is not called, so I think it is fine as is!

**Comment by chatton** [comment]
I added a chain restart in the transfer test, and am bumping callbacks to the new rc5 

cc @colin-axner

File: e2e/tests/transfer/base_test.go [hunk 2]
```diff
@@ -57,6 +57,14 @@ func (s *TransferTestSuite) TestMsgTransfer_Succeeds_Nonincentivized() {
 
 	s.Require().NoError(test.WaitForBlocks(ctx, 1, chainA, chainB), "failed to wait for blocks")
 
+	t.Run("ensure capability module BeginBlock is executed", func(t *testing.T) {
+		// by restarting the chain we ensure that the capability module's BeginBlocker is executed.
+		s.Require().NoError(chainA.StopAllNodes(ctx))
+		s.Require().NoError(chainA.StartAllNodes(ctx))
+		s.Require().NoError(test.WaitForBlocks(ctx, 5, chainA), "failed to wait for blocks")
+		s.InitGRPCClients(chainA)
+	})
```
[the remark this answers is not in the document]
Comment (colin-axner) [reply e2e/tests/transfer/base_test.go:66]: sounds good to me

[the remark this answers is not in the document]
Comment (chatton) [reply e2e/tests/transfer/base_test.go:66]: think the naming is a bit misleading with `base_test.go`, this is actually just happening in the `TestMsgTransfer_Succeeds_Nonincentivized`. Happy to move it to its own separate test though.

**Review by colin-axner — APPROVED** [review]
Thank you @chatton!

**Comment by muku314115** [comment]
Is this issue still up for development?
If yes then would it make sense to just shift restart test to capability_test or genesis_state and use it wherever required? Seems fairly doable

**Pull request merged**